api/v1alpha1/common_types.go

```diff
@@ -113,3 +121,60 @@ type Status struct {
 	// resources without needing to understand resource-specific status details.
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
 }
+
+// HasFinished returns true in the following cases:
+//
+// * If ConditionProgressing is false with ReasonFinished.
+//
+// * If ConditionPreBackupPodsReady is false with any of the "failed" reasons.
+func (in *Status) HasFinished() bool {
+	preBackupCond := meta.FindStatusCondition(in.Conditions, ConditionPreBackupPodsReady.String())
+	// Failed pre backups also count as finished
+	if preBackupCond != nil && isPreBackupFailed(*preBackupCond) {
+		return preBackupCond.Status == metav1.ConditionFalse
+	}
+	progressingCond := meta.FindStatusCondition(in.Conditions, ConditionProgressing.String())
+	if progressingCond != nil && matchAnyReason(*progressingCond, ReasonFinished) {
+		return progressingCond.Status == metav1.ConditionFalse
+	}
+	return false
+}
+
+// HasStarted returns true if ConditionProgressing is true with ReasonStarted, false otherwise.
+func (in Status) HasStarted() bool {
+	condition := meta.FindStatusCondition(in.Conditions, ConditionProgressing.String())
+	if condition != nil && matchAnyReason(*condition, ReasonStarted) {
+		return condition.Status == metav1.ConditionTrue
+	}
+	return false
+}
+
+// IsWaitingForPreBackup returns true if the ConditionPreBackupPodsReady is Unknown with ReasonWaiting, false otherwise.
+func (in Status) IsWaitingForPreBackup() bool {
+	condition := meta.FindStatusCondition(in.Conditions, ConditionPreBackupPodsReady.String())
+	if condition != nil && matchAnyReason(*condition, ReasonWaiting) {
+		return condition.Status == metav1.ConditionUnknown
+	}
+	return false
+}
+
+func isPreBackupFailed(condition metav1.Condition) bool {
+	if matchAnyReason(condition,
+		ReasonSucceeded,
+		ReasonWaiting,
+		ReasonNoPreBackupPodsFound,
+		ReasonReady) {
+		return false
+	}
+	// All other reasons are failures
+	return true
+}
+
+func matchAnyReason(condition metav1.Condition, reasons ...ConditionReason) bool {
+	for _, reason := range reasons {
+		if condition.Reason == reason.String() {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
I wonder if I should move the functions `SetStarted`, `SetFinished` also to the Status struct? Or at least the bits that set the conditions? We would have "getter" and "setter" in one place.

Reply:
there, it's done. WDYT?

Reply:
You will have to call these via `c.Obj().GetStatus()` then. I think the argument is valid, but I don't think that the functions are on the wrong object either. The logic for me goes that "I want to update the _Resource_, so that it's status indicates that the corresponding action was successful". We have this method on the _Resource_ now, which seems perfectly fine to, as _Status_ is just an implementation detail of the _Resource_. On the other hand, it would also be fine in _Status_, as it's the actual implementation.

For now, I'd leave them where they are, because they are used all over the place.

Reply:
FYI: I didn't see your "it's done" when I typed my comment.

Reply:
Ok. So, disregard my lengthy comment https://github.com/vshn/k8up/pull/384#discussion_r587652912. It's a nice and clean implementation now. 👍